{
    "id": "SOLR-11479",
    "title": "Collections API: ADDREPLICA fails if you try to specify property.coreNodeName=foo",
    "details": {
        "labels": "",
        "priority": "Major",
        "components": [],
        "type": "Bug",
        "fix_versions": [],
        "affect_versions": "None",
        "resolution": "Unresolved",
        "status": "Open"
    },
    "description": "if you attempt to specify a property.coreNodeName param when using the Collection API's action=ADDREPLICA the request will fail \u2013 the underlying cause of the failure (in the logs for the node where the underlying Core Admin CREATECORE request is routed) seems like a perplexing chicken/egg problem...\n\n(the various names in the error below are from a test patch i'll attach shortly)\n\n\n   [junit4]   2> Caused by: org.apache.solr.common.SolrException: Unable to create core [solrj_replica_explicit_coreNodeName_shard1_replica_n5]\n   [junit4]   2> \tat org.apache.solr.core.CoreContainer.createFromDescriptor(CoreContainer.java:1049)\n   [junit4]   2> \tat org.apache.solr.core.CoreContainer.create(CoreContainer.java:947)\n   [junit4]   2> \t... 34 more\n   [junit4]   2> Caused by: org.apache.solr.common.SolrException: coreNodeName user_assigned_node_name does not exist in shard shard1: DocCollection(solrj_replica_explicit_coreNodeName//collections/solrj_replica_explicit_coreNodeName/state.json/7)={\n...\n   [junit4]   2> 11157 INFO  (qtp1977346252-47) [n:127.0.0.1:51786_solr c:solrj_replica_explicit_coreNodeName s:shard1 r:user_assigned_node_name x:solrj_replica_explicit_coreNodeName_shard1_replica_n5] o.a.s.s.HttpSolrCall [admin] webapp=null path=/admin/cores params={qt=/admin/cores&coreNodeName=core_node6&collection.configName=conf&name=solrj_replica_explicit_coreNodeName_shard1_replica_n5&property.coreNodeName=user_assigned_node_name&action=CREATE&collection=solrj_replica_explicit_coreNodeName&shard=shard1&wt=javabin&version=2&replicaType=NRT} status=400 QTime=3009\n\n\n\n\n\n...the CREATE core action is failing because the cloud shard we want to ADDREPLICA to says that a replica with that coreNodeName doesn't exist????",
    "attachments": {
        "SOLR-11479.patch": "https://issues.apache.org/jira/secure/attachment/12891769/SOLR-11479.patch"
    },
    "issue_links": {},
    "comments": [
        {
            "date": "2017-10-12T18:22:23+0000",
            "content": "here's a (test only) patch demonstrating the problem ... really straight forward addition to CollectionsAPISolrJTest...\n\n\n  @Test\n  public void testAddReplicaWithCoreNodeNameProp() throws Exception {\n\n    final String collectionName = \"solrj_replica_explicit_coreNodeName\";\n    CollectionAdminRequest.createCollection(collectionName, \"conf\", 1, 2)\n      .process(cluster.getSolrClient());\n\n    final String coreNodeName = \"user_assigned_node_name\";\n    \n    CollectionAdminResponse response = CollectionAdminRequest.addReplicaToShard(collectionName, \"shard1\")\n      .withProperty(CoreDescriptor.CORE_NODE_NAME, coreNodeName)\n      .process(cluster.getSolrClient());\n    \n    assertEquals(0, response.getStatus());\n    assertTrue(response.isSuccess());\n    \n    Replica newReplica = grabNewReplica(response, getCollectionState(collectionName));\n    assertTrue(newReplica.getName().equals(coreNodeName));\n  }\n\n\n ",
            "author": "Hoss Man",
            "id": "comment-16202391"
        },
        {
            "date": "2017-10-12T18:28:30+0000",
            "content": "discovered while trying to fix the tests in SOLR-11469 ",
            "author": "Hoss Man",
            "id": "comment-16202400"
        }
    ]
}